Adding snapshot command to generate rosinstall with SHAs for current …

rsinnet opened the pull request:
I've added a useful feature for capturing a snapshot of the current workspace. It looks like someone had had the idea of implementing this feature as I found an incomplete implementation of a 'snapshot' command. I'm happy to add additional features to this command if desired.

wjwwood commented on src/wstool/config.py:
```diff
@@ -241,14 +241,20 @@ def get_base_path(self):
-    def get_source(self):
+    def get_source(self, versioned=False, vcs_only=False):
         """
         :returns: all elements that got added by user keystrokes
```
Can you please update the doc block here?

rsinnet replied:
Will do. Thanks

wjwwood commented on src/wstool/config_yaml.py:
```diff
@@ -249,7 +249,7 @@ def get_legacy_type(self):
-    def get_legacy_yaml(self):
+    def get_legacy_yaml(self, curr_revision=False):
         """
         return something like
```
Also here.

wjwwood commented on src/wstool/config_yaml.py:
```diff
@@ -378,29 +382,39 @@ def get_path_spec_from_yaml(yaml_dict):
+                 for x in config.get_source(curr_revision, vcs_only)]
+
+    if not items:
+        return
```
I think this causes the behavior to change. Before, when you reached this case, the file was already opened and the header was already written. Now the file will not have been opened and the header will not have been written. I don't know which behavior is correct off-hand, but this is definitely different from before. It might be easier to avoid restructuring this logic to make the change as minimal as possible.

wjwwood replied:
Sure, but in the case of `items` being empty, I think it would be more like it was if the header was written to stdout when `filename` is `None`. Since the header was always written before, even if the `items` was empty. So you can restructure the logic if you like, but in my opinion you either need to restructure this block to write the header before `return` or justify why this new behavior is ok.

rsinnet replied:
You're right! I didn't realize what you meant at first. I've fixed it so the header will be written even if `items` is empty. Thanks.

wjwwood commented on src/wstool/config_yaml.py:
```diff
@@ -378,29 +382,39 @@ def get_path_spec_from_yaml(yaml_dict):
+    else:
+        content += yaml.safe_dump(items)
+
+    if filename:
```
This also changes the behavior of how filename is interpreted. I think this is required based on how the snapshot tool works, but you should add a line to the doc explaining this dual purpose of filename.

rsinnet replied:
My long comment above addresses this. This is only necessary if we wish to have the ability to write to _stdout_ instead of only being able to write to a file.

rsinnet commented:
Thanks for the comments. I'll address the issues you've pointed out and resubmit.

rsinnet commented:
I'll also fix the build failures.

rsinnet commented:
I added five tests to cover the new feature. I think all the issues mentioned above are now fixed. Let me know if anything else needs to be modified. Thanks.

tfoote commented:
Thanks this looks very useful. I was actually looking for this just the other day. 

I'd suggest following the same usage pattern as `vcs` with `export` as the command and an option `--exact` to give commit references instead of branchnames. The basic version is mostly a pass through to the underlying .rosinstall file assuming it's up to date. However it could check that it's up to date and warn etc. 

I haven't had a chance to read through the code but what's the current behavior if the checkouts are out of date for example the url has changed?

rsinnet commented:
@tfoote I've renamed the `snapshot` command to `export` and added the `--exact` option.

You raise an interesting point: What should the behavior be when the workspace doesn't match the .rosinstall? Currently, it just does a pass-through and takes the information from the .rosinstall but takes the commit hash from the checked out version. I suppose it doesn't make much sense to mix information like this and in some cases could result in unexpected behavior so export should probably take the current URI and version as well. I'll go ahead and make this change since it seems appropriate.

I could also add a `--spec` option that bases the snapshot on the workspace .rosinstall (in the way that the `info` command outputs Spec information). What do you think?

rsinnet commented:
I've addressed my last comment by adding a `--spec` option to the `export` command and implementing reasonable behavior. I've also added a new test suite for the `export` command that demonstrates the correct behavior. Let me know if you think any additional changes are needed or if you think the default behavior of the `export` command should be changed (currently, it takes the version name from the workspace, i.e., not `--spec` and not `--exact`).

@tfoote I haven't added any warnings if checkouts are out of date. Perhaps we can create an issue and discuss which warnings would be desirable.

Finally, I'm not sure why my last push didn't kick off a build, but on my fork I have Travis CI hooked up and everything worked.

tfoote commented:
Attempting to retrigger travis with close and reopen.

tfoote commented:
That sounds like a good compromise where the two options explicitly bracket the default behavior. I'm busy at the moment with a pending deadline. I'll try to get a full read through this week though. 

It looks like Travis passed.Coveralls seems to be hanging continuously.

rsinnet commented:
Thanks again for looking through the changes and the useful comments. I'll try to look more into the coveralls issue this week. So far, I did the following on my machine:

1. Load up travis-python container.
2. Install prerequisites (like mock)
3. Run commands in .travis.yml
4. Run coveralls command

For me, nothing hangs locally and I end up with a coverage file and output that says coverage submitted. Looking in the online logs for the Travis builds for this pull request that nothing hangs there either, although I do not see a comment about coverage being submitted. Is it likely that my changes to the code have caused this or could it be some other issue? I will try submitting a trivial pull request to see if coveralls succeeds.

rsinnet commented:
@tfoote I created an empty pull request (#99) and it seems that coveralls exhibits the same hanging with no changes from master. I suspect there is a configuration issue here and that this hanging is not caused by code changes in the present pull request (#98).

wjwwood commented:
Yeah, looks like coverage reporting is just broken. I'll merge this as-is.
